apple-color-emoji: init at unstable-2020-04-13

btwiusegentoo opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Add Emoji font used by Apple devices(macOS,iOS,iPad OS etc)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
maintainers: add btwiusegentoo

Commit:
apple-color-emoji: init at unstable-2020-04-13

btwiusegentoo commented:
>Emoji fonts (under the fonts subdirectory) are under the SIL Open Font License, version 1.1.
>Tools and some image resources are under the Apache license, version 2.0.
apple-emoji-linux repo itself uses Apache license. but reading above sentence, maybe It's under OFL?

OPNA2608 commented on pkgs/data/fonts/apple-color-emoji/default.nix:
```diff
@@ -0,0 +1,32 @@
+  name = "apple-color-emoji";
+
+  src = fetchurl {
+    url = https://github.com/samuelngs/apple-emoji-linux/releases/download/latest/AppleColorEmoji.ttf;
```
Please quote URLs. See https://github.com/NixOS/rfcs/blob/master/rfcs/0045-deprecate-url-syntax.md
```suggestion
    url = "https://github.com/samuelngs/apple-emoji-linux/releases/download/latest/AppleColorEmoji.ttf";
```

OPNA2608 commented on pkgs/data/fonts/apple-color-emoji/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  meta = with stdenv.lib; {
+    description = "A Color typeface used by iOS and macOS to display emoji";
+    homepage = https://github.com/samuelngs/apple-emoji-linux;
```
Please quote URLs. See https://github.com/NixOS/rfcs/blob/master/rfcs/0045-deprecate-url-syntax.md
```suggestion
    homepage = "https://github.com/samuelngs/apple-emoji-linux";
```

OPNA2608 reviewed (changes requested):
Could you try and give building the fonts a shot instead? Any passing commit to upstream's repo will invalidate the sha256 hash until it is fixed by someone again. Check their readme for build instructions.

OPNA2608 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -27241,4 +27241,6 @@ in
 
   navidrome = callPackage ../servers/misc/navidrome {};
+
+  apple-color-emoji = callPackage ../data/fonts/apple-color-emoji {};
```
Not an issue per se, but all-packages.nix once had a rough structure & sorting. I'd personally move it to the start of `### DATA` to the bulk of other `/data/*` stuff, sorted alphabetically. But this structure isn't required from what I know.

btwiusegentoo commented:
Thanks for all suggestions. I rewrited to pull from repo and build. I used a specific commit(from 2020-04-13) because latest couldn't build successful. Also, I quoted urls and moved the line in all-packages.nix to ### DATA sorted alphabetically.
And I was looking other packages and I found that can select multiple licenses so I added asl20.
tested with nix-build, installing system-wide.

OPNA2608 commented on pkgs/data/fonts/apple-color-emoji/default.nix:
```diff
@@ -0,0 +1,29 @@
+  enableParallelBuilding = true;
+
+  installPhase = ''
+    install -m444 -Dt $out/share/fonts/apple-color-emoji AppleColorEmoji.ttf 
```
The Makefile has an install method you can reuse, just need to patch the install path.
```suggestion
    substituteInPlace Makefile --replace 'HOME/.local/' 'out/'
```

OPNA2608 reviewed (changes requested):
Getting there!

You should import the needed dependencies right away instead of the whole `pkgs` set, this'll make it much easier to override specific dependencies and is generally how packages are written. You can also skip the `configurePhase` with `dontConfigure` if there's nothing to be configured.

```nix
{ stdenv, lib, fetchFromGitHub
, python3, optipng, zopfli, pngquant, imagemagick, which
}:

let
  python = python3.withPackages (ps: with ps; [
    fonttools nototools
  ]);
in
stdenv.mkDerivation rec {

  src = fetchFromGitHub {
  };

  dontConfigure = true;

  nativeBuildInputs = [ python optipng pngquant imagemagick which ];
}
```
etcetc.

There's also a code formatter specifically for nixpkgs - `nixpkgs-fmt`. Might be a good practice to run it on your derivation after these changes, I should start doing this more too.

Lastly, please squash the commits down and make sure they follow the contribution guidelines. Then it's all OK in my eyes.

```
maintainers: add btwiusegentoo
apple-color-emoji: init at unstable-2020-04-13
```

btwiusegentoo commented:
Thanks! I merged the changes and formatted with nixpkgs-fmt, finally runned `git rebase` to fix the commit logs and pushed.

OPNA2608 commented on pkgs/data/fonts/apple-color-emoji/default.nix:
```diff
@@ -0,0 +1,39 @@
+  meta = with stdenv.lib; {
+    description = "Color typeface used by iOS and macOS to display emoji";
+    homepage = "https://github.com/samuelngs/apple-emoji-linux";
+    license = [ licenses.ofl licenses.asl20 ];
```
```suggestion
    license = with licenses; [ ofl asl20 ];
```

OPNA2608 reviewed (changes requested):
_One very last thing_, then it's done for good IMO.

btwiusegentoo commented:
ok,done!

Infinisil commented on maintainers/maintainer-list.nix:
```diff
@@ -9203,4 +9203,10 @@
     github = "deifactor";
     githubId = 30192992;
   };
+  btwiusegentoo = {
```
Got a conflict because of this here, I suggest putting this in the alphabetically correct place instead because of this.

btwiusegentoo replied:
Thanks for the suggestion. I fixed, rebased onto latest upstream and pushed

jtojnar commented on pkgs/data/fonts/apple-color-emoji/default.nix:
```diff
@@ -0,0 +1,39 @@
+  enableParallelBuilding = true;
+
+  patchPhase = ''
+    substituteInPlace Makefile --replace 'HOME/.local/' 'out/'
```
With [this patch](https://github.com/samuelngs/apple-emoji-linux/pull/7), you can just use `makeFlags = [ "PREFIX=${placeholder "out"}" ];`

btwiusegentoo replied:
Thanks! So, should I wait for PR merge and make changes after that?

Infinisil replied:
PR is merged :)

btwiusegentoo replied:
Thanks for reminding me! bad news, I changed to latest rev, tried to build. still it failes to build with this error.
Same error as [GitHub CI](https://github.com/samuelngs/apple-emoji-linux/runs/1036473672?check_suite_focus=true)
```
Dropped outline ('glyf', 'CFF ') and related tables.
Output font 'AppleColorEmoji.ttf' generated.
Traceback (most recent call last):
  File "map_pua_emoji.py", line 73, in <module>
    main(sys.argv)
  File "map_pua_emoji.py", line 69, in main
    add_pua_cmap(argv[1], argv[2])
  File "map_pua_emoji.py", line 61, in add_pua_cmap
    glyph_name = get_glyph_name_from_gsub([ch1, ch2], font)
  File "map_pua_emoji.py", line 43, in get_glyph_name_from_gsub
    ligatures = lookup.SubTable[0].ligatures
  File "/nix/store/pf57mqvrygkxl8r4s75f2fd7qff394ji-python3-3.8.5-env/lib/python3.8/site-packages/fontTools/ttLib/tables/otBase.py", line 548, in __getattr__
    raise AttributeError(attr)
AttributeError: ligatures
make: *** [Makefile:145: AppleColorEmoji.ttf] Error 1
rm AppleColorEmoji.tmpl.ttx build/emoji/emoji_ufe82b.png
builder for '/nix/store/n7mhpi8v4lvpizca88iksi02144ixc8l-apple-color-emoji-unstable-2020-08-27.drv' failed with exit code 2
error: build of '/nix/store/n7mhpi8v4lvpizca88iksi02144ixc8l-apple-color-emoji-unstable-2020-08-27.drv' failed
```
I would like to make a PR to fix if I could, but I'm new to programming in general. I don't know where to start investigating the reason.
Seems like build error starts after [updating emoji icons](https://github.com/samuelngs/apple-emoji-linux/commit/35656af89ce7e1e4b2afc9360eaa32bac2a0d705)

sternenseemann replied:
I think the issue is that the toolchain relies on outdated versions of fonttools and nototools. The versions we ship in nixpkgs don't even support python 2 anymore which apple-emoji-linux seems to require.

What we could try is to reuse the noto-fonts-emoji toolchain to build apple-color-emoji, similarly to how twitter-color-emoji works in nixpkgs.

Infinisil commented:
There doesn't seem to be much interest in this PR anymore, I'll close it for now.
